core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -878,15 +996,41 @@ class KafkaController(val config: KafkaConfig,
    * is complete (i.e. there is no other partition with a reassignment in progress), the reassign_partitions znode
    * is deleted.
    *
-   * `ControllerContext.partitionsBeingReassigned` must be populated with all partitions being reassigned before this
+   * `ControllerContext.partitionsBeingReassigned` must be populated with all the zk-persisted partition reassignments before this
    * method is invoked to avoid premature deletion of the `reassign_partitions` znode.
    */
   private def removePartitionsFromReassignedPartitions(partitionsToBeRemoved: Set[TopicPartition]) {
     partitionsToBeRemoved.map(controllerContext.partitionsBeingReassigned).foreach { reassignContext =>
       reassignContext.unregisterReassignIsrChangeHandler(zkClient)
     }
 
-    val updatedPartitionsBeingReassigned = controllerContext.partitionsBeingReassigned -- partitionsToBeRemoved
+    removePartitionsFromZkReassignment(partitionsToBeRemoved)
+
+    controllerContext.partitionsBeingReassigned --= partitionsToBeRemoved
+  }
+
+  private def removePartitionFromReassignedPartitions(partitionToBeRemoved: TopicPartition) {
+    controllerContext.partitionsBeingReassigned.get(partitionToBeRemoved) match {
+      case Some(reassignContext) =>
+        reassignContext.unregisterReassignIsrChangeHandler(zkClient)
+
+        if (reassignContext.persistedInZk) {
+          removePartitionsFromZkReassignment(Set(partitionToBeRemoved))
+        }
+
+        controllerContext.partitionsBeingReassigned.remove(partitionToBeRemoved)
+      case None =>
+        throw new IllegalStateException("Cannot remove a reassigning partition because it is not present in memory")
+    }
+  }
+
+  private def removePartitionsFromZkReassignment(partitionsToBeRemoved: Set[TopicPartition]): Unit = {
+    if (!zkClient.reassignPartitionsInProgress()) {
+      debug(s"Cannot remove partitions $partitionsToBeRemoved from ZooKeeper because there is no ZooKeeper reassignment present")
```

Comment:
Come to think of it this would get called on API every reassignment so it may be worth to not log it?

Reply:
I think it's OK to keep the log.